### DISCUSSION TITLE
ASCII是什麼 如何比較字串與數字

### DISCUSSION THREAD

**Discussion by xCk27x** [1101-GS4719-C/Python-Community-GS4719-C#177] [Show and tell]
**ASCII碼** 簡單來說 就是一些常見字元(char)會對應從 0~127 的十進位數字
但Python中並沒有字元 所以這裡說的字元在Python中即代表 \*長度為**1**的字串\* ex : 'A', '7', '\n'

而要取出一個字元的ASCII碼則需要用到 ord() 這個函式
例如 '0' 對應到的ASCII碼是 48 
所以執行 ord('0') 會得到 48 的結果
而要將數字轉換成對應的字元則需要使用 chr() 
例如執行 chr(48) 會得到 '0'
![螢幕擷取畫面 (18)](https://user-images.githubusercontent.com/90547641/144186237-d84c3021-9152-4e0b-9426-61091348acef.png)

注意到 ord() 輸出的型態為 **int** 所以可以跟其他數字進行比較 
所以執行 ord('0') == 48 會得到 True
同理 執行 chr(48) == '0' 也會得到 True
![螢幕擷取畫面 (20)](https://user-images.githubusercontent.com/90547641/144188557-81ac59cf-ece8-4a0e-b3bc-eb3cd38096b1.png)

然而 你可能會想說沒事幹嘛把字元轉成數字 這有什麼好處
事實上 這在進行字元之間的轉換時非常有用
例如 str1 = 'a' 
今天我想把 str1 改成 'A'
除了利用 str1.upper() 以外
我們也可以利用以下程式進行轉換
str1 = chr(ord(str1) - 32)
原因是在ASCII中 小寫字母是大寫字母加32
![螢幕擷取畫面 (22)](https://user-images.githubusercontent.com/90547641/144241304-3d299d9a-1519-4e84-bfc4-d9fffe9ad75b.png)
因此我們就可以得到大寫的 'A' 了

另一種應用是將字元變成數字加以比較 舉例來說
我請使用者輸入一個符號 而我想知道他輸入的是不是英文字母
便能定義一個叫 is_aplhabet 函式如圖
![螢幕擷取畫面 (25)](https://user-images.githubusercontent.com/90547641/144249215-05125a09-ef7d-4e4a-9fed-1b94b7b1bc0d.png)
再細一點甚至能區分大小寫或是某個區間
p.s. 直接寫if('a'<=char<='z' or 'A'<=char<='Z'): 
      也是可以 但電腦在比較時仍是比較其對應的ASCII值
      所以這樣寫比較能搞懂電腦是怎麼比較字元的

事實上字串其實也可以互相比較 之後說不定會再寫一篇 有興趣可以試試看

**[#這裡是ASCII碼的對照表#](https://onepc.cc/ascii)** 可以參考一下喔

**Comment by xCk27x** [comment]
順帶一提 ASCII中的0~31以及127所代表的是控制字元
例如 換行字元 '\n' 對應10
以及 刪除 DEL 對應127
他們出現時會對電腦做出各種不同的控制

**Comment by enoch-lightning** [comment]
寫一個is_aplhabet函式，使用只要是小寫字母，或大寫字母，就會得到True=真=對，而使用print()函式輸出結果至瑩幕還蠻有趣的。

**Comment by Jie2222** [comment]
ASCII碼給予了不同電腦之間相同的語言，使彼此之間能夠共享檔案或文件

**Comment by xCk27x** [comment]
關於字串的比較，其實就是從第一個字符依序比較，若第一個字符相同則繼續比較第二個字符，直到出現兩個不同的字符

**Comment by xCk27x** [comment]
但我其實蠻好奇字串中的'中文'是否能換成ASCII碼 不知道有沒有人能解答

**Reply by HSU-YU-JEN** [reply]
有看到網路上有類似的內容~
https://social.msdn.microsoft.com/Forums/zh-TW/35d324a7-df8b-4980-89c0-b50f0205b4d7/288582016040636vbnet3034020013259912338336681ascii652922087736681?forum=232

**Reply by HSU-YU-JEN** [reply]
還有這篇也可以參考，補發一下
https://social.msdn.microsoft.com/Forums/zh-TW/6f0d7b19-bd41-4e1a-ac2f-51167941d0f9/24819355312183920013259912338336319ascii?forum=232

**Comment by chupochun** [comment]
以前就有聽過ASCII碼謝謝版主的介紹，讓我更加認識它了

**Comment by ghost** [comment]
對於資電之類的領域，這可以說是非常需要了解的知識RR

**Comment by Yuhsin000** [comment]
感覺是很好的轉換工具!
